src/commander.ts

```diff
@@ -3,126 +3,56 @@ import {Job} from "./job";
-        parser.getJobs().forEach(j => whenPadEnd = Math.max(j.when.length, whenPadEnd));
-
-        let stagePadEnd = 0;
-        parser.getStageNames().forEach(s => stagePadEnd = Math.max(s.length, stagePadEnd));
-
-        let descriptionPadEnd = 0;
-        parser.getJobs().forEach(j => descriptionPadEnd = Math.max(j.description.length, descriptionPadEnd));
-
-        const neverJobs = jobs.filter(j => j.when === "never");
-        const nonNeverJobs = jobs.filter(j => j.when !== "never");
-
-        const renderLine = (job: Job) => {
-            const needs = job.needs;
-            const allowFailure = job.allowFailure ? chalk`{black.bgYellowBright  ONLY WARN }` : chalk`{black.bgRed  CAN FAIL  }`;
-            let jobLine = `${job.getJobNameString()}  ${job.description.padEnd(descriptionPadEnd)}`;
-            jobLine += chalk`  {yellow ${job.stage.padEnd(stagePadEnd)}}  ${job.when.padEnd(whenPadEnd)}  ${allowFailure.padEnd(11)}`;
-            if (needs) {
-                jobLine += chalk`  [{blueBright ${needs.join(",")}}]`;
-            }
-            writeStreams.stdout(`${jobLine}\n`);
-        };
-
-        neverJobs.forEach((job) => renderLine(job));
-        nonNeverJobs.forEach((job) => renderLine(job));
+        await JobExecutor.runLoop(jobs, stages, potentialStarters, privileged);
+        await Commander.printReport(writeStreams, jobs, stages, parser.jobNamePad);
     }
 
-    static async runSingleJob(parser: Parser, writeStreams: WriteStreams, jobName: string, needs: boolean, privileged: boolean) {
-        const jobs: Job[] = [];
-        const foundJob = parser.getJobByName(jobName);
-        jobs.push(foundJob);
+    static async runSingleJob(parser: Parser, writeStreams: WriteStreams, jobArgs: string[], needs: boolean, privileged: boolean) {
+        const jobs = parser.jobs;
+        const stages = parser.stages;
 
+        let jobPoolMap: Map<string, Job>;
+        let potentialStarters: Job[] = [];
         if (needs) {
-            // Recursive backwards traversal to find parent needs.
-            let needed: string[] = [];
-            if (foundJob.needs) {
-                needed = needed.concat(foundJob.needs);
-            }
-            while (needed.length > 0) {
-                const need = needed.pop();
-                if (need) {
-                    const needJob = parser.getJobByName(need);
-                    jobs.unshift(needJob);
-                    if (needJob.needs) {
-                        needed = needed.concat(needJob.needs);
-                    }
-                }
-            }
-        }
-
-        for (const job of jobs) {
-            await job.start(privileged);
+            jobPoolMap = new Map<string, Job>(jobs);
+            jobArgs.forEach(jobName => {
+                const job = Utils.getJobByName(jobs, jobName);
+                jobPoolMap.set(jobName, job);
+                potentialStarters = potentialStarters.concat(JobExecutor.getPastToWaitFor(jobs, stages, job));
```

Comment:
Move if needs to this line only to depublicate

Reply:
Job pool map is also special